### PR TITLE
KAFKA-2434: Remove identical topic subscription constraint for roundrobin strategy in old consumer API

### PR DESCRIPTION
**Pull request by noslowerdna**

### CODE REVIEW

**Comment by noslowerdna** [comment]
Not sure why the @asfbot build failed. There aren't any obvious test failures in the build log, and I've built the project successfully locally several times.

**Comment by noslowerdna** [comment]
@junrao Would you be able to review this pull request prior to the 0.10 release?

**Comment by junrao** [comment]
@onurkaraman : Do you want to take a look? @noslowerdna : This is probably not an 0.10.0 blocker since we now recommend people to use the new java consumer, which doesn't have that constraint.

**Comment by granthenke** [comment]
@junrao Do we recommend that yet? I think it was decided not to remove the beta tag for 0.10 on the consumer. ([KAFKA-3283](https://issues.apache.org/jira/browse/KAFKA-3283))

**Comment by junrao** [comment]
@granthenke : The issue listed there (KAFKA-3177) seems rare. Given that we have cleaned the new consumer api, my feeling is that the new consumer is ready for production usage.

**Comment by onurkaraman** [comment]
@junrao sure. I'll take a look at it.

File: core/src/main/scala/kafka/consumer/PartitionAssignor.scala [hunk 1]
```diff
@@ -77,18 +72,11 @@ class RoundRobinAssignor() extends PartitionAssignor with Logging {
...
-      // check conditions (a) and (b)
-      val (headTopic, headThreadIdSet) = (ctx.consumersForTopic.head._1, ctx.consumersForTopic.head._2.toSet)
-      ctx.consumersForTopic.foreach { case (topic, threadIds) =>
-        val threadIdSet = threadIds.toSet
-        require(threadIdSet == headThreadIdSet,
-          "Round-robin assignment is allowed only if all consumers in the group subscribe to the same topics, " +
-            "AND if the stream counts across topics are identical for a given consumer instance.\n" +
-            "Topic %s has the following available consumer streams: %s\n".format(topic, threadIdSet) +
-            "Topic %s has the following available consumer streams: %s\n".format(headTopic, headThreadIdSet))
+      val allThreadIds = ctx.consumersForTopic.flatMap { case (topic, threadIds) =>
```
Comment (onurkaraman) [inline core/src/main/scala/kafka/consumer/PartitionAssignor.scala]: A `ConsumerThreadId` can be interested in multiple topics, so `allThreadIds` could have duplicate `ConsumerThreadId`s. Do we want to deduplicate and sort?

Comment (noslowerdna) [inline core/src/main/scala/kafka/consumer/PartitionAssignor.scala]: @onurkaraman There are duplicates, but they are grouped into separate lists -- `allThreadIds` is a list of lists. Each nested list of `ConsumerThreadId`s would remain sorted according to how `zkUtils.getConsumersPerTopic(...)` returns it. We need to maintain that organization as below when assigning each topic partition we first exclude the `ConsumerThreadId` lists from `allThreadIds` where no member of the list is interested in the topic.

Comment (onurkaraman) [inline core/src/main/scala/kafka/consumer/PartitionAssignor.scala]: Just to make sure we're on the same page: are you suggesting `allThreadIds` is literally of type `List[List[ConsumerThreadId]]`, or just figuratively speaking (as in, there's just one list whose contents are as though you poured in all of one list's elements, then the next list's elements, etc)? The flatMap makes `allThreadIds` of type `Iterable[ConsumerThreadId]`.

Comment (noslowerdna) [inline core/src/main/scala/kafka/consumer/PartitionAssignor.scala]: My apologies, I wrote that code a while ago and overlooked that it is indeed a flatMap. Looking at the new consumer round-robin implementation, for consistency's sake I agree that we should de-dup and sort the consumers. I'll make the necessary changes and update the pull request. Thanks for catching that.

**Comment by onurkaraman** [comment]
LGTM but can we get Jenkins to pass?

**Comment by noslowerdna** [comment]
retest this please

**Comment by noslowerdna** [comment]
Started a new Jenkins build.

**Comment by noslowerdna** [comment]
Failed on checkstyle (unused import in an unchanged class, was [recently fixed](https://github.com/apache/kafka/commit/d6b450ced57bb5ab8a7164358e6a360f97a7b5cc)), will rebase off trunk, squash commits, and try another build.

**Comment by noslowerdna** [comment]
retest this please

**Comment by guozhangwang** [comment]
@noslowerdna could you rebase?

**Comment by noslowerdna** [comment]
@guozhangwang rebased.

**Comment by guozhangwang** [comment]
Merged to trunk.

**Pull request closed without merging**